### PR TITLE
Support new window function `ntile`

### PR DESCRIPTION
**Pull request by lgbo-ustc**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...
Support window function `ntile`.
```
insert into test_data values(1,2), (1,3), (1,4), (2,5),(2,6);
select a, b, ntile(2) over (partition by a order by b rows between unbounded preceding and unbounded following ) from test_data;

┌─a─┬─b─┬─ntile(2) OVER (PARTITION BY a ORDER BY b ASC)─┐
│ 1 │ 2 │                                             1 │
│ 1 │ 3 │                                             1 │
│ 1 │ 4 │                                             2 │
│ 2 │ 5 │                                             1 │
│ 2 │ 6 │                                             2 │
└───┴───┴───────────────────────────────────────────────┘
```
### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by lgbo-ustc** [comment]
There is a related issue, https://github.com/ClickHouse/ClickHouse/issues/28989

File: tests/queries/0_stateless/01591_window_functions.sql [hunk 1]
```diff
@@ -545,3 +545,7 @@ select count() over (w) from numbers(1) window w as ();
 
 -- nonexistent parent window
 select count() over (w2 rows unbounded preceding); -- { serverError 36 }
+
+-- ntile
```
Comment (vdimir) [inline tests/queries/0_stateless/01591_window_functions.sql:549]: I'd suggest create separate test file for this function and cover more cases (different arg types, corner cases with illegal arguments and unsupported frames, etc)

Comment (vdimir) [reply tests/queries/0_stateless/01591_window_functions.sql:549]: Result with Float argument unexpected:
```sql
select
	number % 2 as a,
	number as b,
	ntile(2 :: Float64) over (partition by a order by b)  as ntile
from numbers(10);


┌─a─┬─b─┬─ntile─┐
│ 0 │ 0 │     1 │
│ 0 │ 2 │     2 │
│ 0 │ 4 │     3 │
│ 0 │ 6 │     4 │
│ 0 │ 8 │     5 │
│ 1 │ 1 │     1 │
│ 1 │ 3 │     2 │
│ 1 │ 5 │     3 │
│ 1 │ 7 │     4 │
│ 1 │ 9 │     5 │
└───┴───┴───────┘

```

Same for string: `ntile('2')`

Comment (vdimir) [reply tests/queries/0_stateless/01591_window_functions.sql:549]: Non constant argument probably should be forbidden also
```sql
select number%2, number, ntile(number + 2) over (partition by number % 2 order by number)  as ntile from numbers(10);
```

File: src/Processors/Transforms/WindowTransform.cpp [hunk 2]
```diff
@@ -1970,6 +1969,121 @@ struct WindowFunctionRowNumber final : public WindowFunction
...
+            const auto & current_block = transform->blockAt(transform->current_row);
+            const auto & workspace = transform->workspaces[function_index];
+            auto n = (*current_block.input_columns[
+                    workspace.argument_column_indices[0]])[
+                        transform->current_row.row].get<Int64>();
+            if (n <= 0)
+            {
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "ntile's argument must > 0");
+            }
+            buckets = static_cast<UInt64>(n);
```
Comment (vdimir) [inline src/Processors/Transforms/WindowTransform.cpp:2002]: If we cant to `UInt64`, let's forbid bigger argument types (`(U)Int128/256`), `Decimals`, and `Float`. Or better just allow only `(U)Int8/16/32/64`

File: src/Processors/Transforms/WindowTransform.cpp [hunk 3]
```diff
@@ -1970,6 +1969,121 @@ struct WindowFunctionRowNumber final : public WindowFunction
...
+        UInt64 bucket_num = 1;
+        for (UInt64 r = 0; r < current_partition_rows;)
+        {
+            auto current_bucket_capacity = bucket_capacity;
+            if (capacity_diff > 0)
+            {
+                current_bucket_capacity += 1;
+                capacity_diff--;
+            }
+            Int64 left_rows = static_cast<Int64>(current_bucket_capacity);
```
Comment (vdimir) [inline src/Processors/Transforms/WindowTransform.cpp:2043]: Why do we cast to signed?

**Review by vdimir — APPROVED** [review]
Thank you for making an effort to support this window function!

**Comment by alexey-milovidov** [comment]
#46763

**Pull request merged**